Initial stab at new impl using an eavs index for datoms

rundis opened the pull request:
First of: This PR is obviously not ready for merging, but it I guess it contains enough to make some decisions on whether this is a path worth pursuing :)

A few highlights:
- Testing (using [criterium](https://github.com/hugoduncan/criterium) on my mac (tm) ) indicates a performance improvement of roughly 3X for `transact-sources!`   
- The `datoms` ns is very much inspired (/copied in large parts) from datascript. Not well-versed enough in license stuff to assess whether that's ok or not...
- Some of the optimisations might be useful for the existing implementation, regardless of whether the eavs track pans out or not
- Many of the changes has been measured with/without transients and various combinations of using reduce/loop-recur/transducers and whatnot. It feels like it's not that much more to be squeezed out without rethinking the algorithm/approach altogether. 

 lmk if this is/could become useful, and if so how you wish to proceed.

rundis committed:
Initial stab at new impl using an eavs index for datoms

magnars commented:
Very cool! 👏👏👏 Looking forward to having you explain it to me IRL. :-)

I had planned something like this myself, but then I anticipated more indices. Have you looked into using other indicies than `eavs` and found them not useful, or is that something for the future?

rundis committed:
Speed up conflicting values/sources checking

rundis committed:
Fix regression with not retracting obsolete datoms from eavs

rundis committed:
Make source datoms replace helper faster

- Also minor optimization for entity-ref lookups (at the cost of potentially
missing some entity-validation... might not be worth it)

rundis committed:
Improve perf for conflict checks, refs/tx-data creation

- Merge conflicting values and conflicting sources checks
- Temporarily "store" diffs per source and calculate tx-data by massaging the
  accumulation of these
- Faster lookup-refs creation by using a loop rather than transducer

rundis commented:
Ok, so by now I don't see how I may squeeze much more out of the diffing performance. (on account of readability/maintenance I might even have taken a bit too far :)). Now `explode` (of entities into datoms) accounts for more than 50% of time spent for the test-scenarios I have run.

Regarding additional indexes, I did consider an ave(s) index for lookup-refs (event wasted time implementing a version using it... jeez), but beating a plain ole' hash-map for performance wasn't even a close call :)   

Anyways it appears to now be between 5 - 20 times faster than the current implementation, depending on use case (based on criterium tests on my machine that is)

magnars commented:
That's an impressive piece of work. Well done. :-) 

Since we are breaking backwards compatibility, let's keep this in a new namespace, and update the docs accordingly. In particular, I think we should remove/not implement:

- using :db/id to identify entities
- partial updates

Apart from these, what is the status with regards to feature parity with the old implementation? Do the old (non implementation-specific) tests all run?

rundis committed:
Add transact! improve/add tests and fix a bug as a result

rundis committed:
Remove attrs from db to avoid making it potential future breakage

rundis commented:
tx

## API
Added `transact!` so now it should have all of the functions announced as public in the readme (bar the db/id and partial stuff which I didn't implement as it seemed highly discouraged)

Note: __I´ve spotted instances in the wild where functions not announced as public are used, most notably `with` and `explode`. But hey they did that at their own risk__ 😁 

## Tests
> Do the old (non implementation-specific) tests all run?
When starting the refactor I focused on **only** writing tests for the public API. That made the test somewhat more cumbersome to write, but refactoring the internals a lot less painful! 
I've done a run through comparing the old `core_tests` and hopefully the new api tests should now cover most if not all of the relevant test cases from the existing core-test ns.

## Whats more ?
- I haven't looked and the `export` and `report` namespaces, the former should be trivial to update/create new ns for, the latter is probably a bit more work I imagine. 
- Update readme
- Real world testing
- stuffs ...

magnars commented:
> Note: **I´ve spotted instances in the wild where functions not announced as public are used, most notably `with` and `explode`. But hey they did that at their own risk** 😁

I consider both of these a part of the public API, even if the docs were lacking. Isn't `transact!` most easily implemented using `with`? And we still have `explode`, don't we? :-)

> When starting the refactor I focused on **only** writing tests for the public API. That made the test somewhat more cumbersome to write, but refactoring the internals a lot less painful!

That sounds like a good approach. 

> I've done a run through comparing the old `core_tests` and hopefully the new api tests should now cover most if not all of the relevant test cases from the existing core-test ns.

This reminds me that TDD-tests aren't the best regression tests when refactoring internals. I'm not sure what to learn from that tho. 

> * I haven't looked and the `export` and `report` namespaces, the former should be trivial to update/create new ns for, the latter is probably a bit more work I imagine.

The `report` namespace isn't very affected by this change. I think the only difference is in Datoms over `[e a v]` vectors.

rundis committed:
Add explode, with and get-datoms functions to public api

- Opted to provide a custom impl of explode (using the internal
`explode-entity-maps` function) with a simple signature. Also this
way clients don't need to know about the internal `Datom` type at all.
- `get-datoms` exposed as a way to get a complete view of the datoms in the db (as a list
of `[e a v]`)

rundis committed:
Add datom value comparator protocol

By default only comparable values are supported, by extending the
DatomValueComparator protocol, things like maps and sets can be supported as
datom values.

rundis committed:
Add export-db and prune-diffs to core2 namespace

rundis committed:
Refactor common core stuff into a core-helpers ns

rundis commented:
- Added a protocol for custom datom value comparator override
- Opted to add export-db and prune-diffs to `core2` rather than creating an `export2` ns
- Refactored some common stuff between `core` and `core2` into a helper ns. 

Bar further comments on the commits provided so far I guess remaining work now consists of:
- Maybe find a better name for `core2` (alternatively keep it, but introduce an `api` ns "in front" 
- Update readme (maybe best if you do it, to keep writing style to your liking!?)
- (make an effort to ensure it works properly in clojurescript, which It might(/probably will) not after all these changes
- Bump version and publish :)

magnars commented:
Looking good! 👍

Since `core` is legacy "never touch again"-code, would it make sense to keep them entirely separte, ie. no common stuff?

I like the idea of renaming `core2` to `api`, and adding a deprecation notice to the `core`-ns and README. Not "deprecated" as in "going away", but deprecated as in "use that other thing if you can".

I'm happy to make changes to the README. Could you give me a pointer as to what needs to change?

rundis committed:
Minor optimizations for initial diff and small diff with large db

- By not invoking get-entity-ref more than necessary some time can be saved when
initially diffing (aka with empty db) especially when there are many entities
- When there are few db/adds compared to total number of datoms in db, conflict
checking can be optimized by just checking the additions as opposed to looping
through all datoms.

rundis committed:
Rename core2 to api

rundis commented:
- Added a few minor optimisations based on known real-world use cases with large initial dataset followed by small(ish) updates later on. 
- Also `core2` is now `api`

> Not "deprecated" as in "going away", but deprecated as in "use that other thing if you can".

Not sure what best practice is for documenting deprecations in clojure land, so I've left that out 

> Since core is legacy "never touch again"-code, would it make sense to keep them entirely separate, ie. no common stuff?

I opted to put it in common/shared place since `find-attrs` was used by the `export` and `report` namespace. I was much more in doubt about the other functions though. Event though the test-coverage should be decent, I guess the chances of slipups when fiddling with shared stuff in the future is deffo there. Anyways It's your lib so I'll leave the decision up to you 😄 

> I'm happy to make changes to the README. Could you give me a pointer as to what needs to change?
- Any references to `core` should be changed to api
- The `export-db` and `prune-diffs` functions are now part of the api ns, in addition `get-datoms` is also considered part of the api and needs a line or two I suppose
-  db/id and partial updates are gonners
- Maybe a link to the "old" readme? 
- I guess a rationale (a few lines?) for why the change was made is called for


Oh one last thing I forgot, `refs` just keeps accumulating. That's how it works currently I believe. Maybe just maybe we should remove entries from the refs map when all datoms using it are gone from the db?
The only rationale I can think of for not removing, is that at some point in the future the same lookup ref is again asserted you reuse the eid thus less "holes". But it doesn't sound entirely convincing a reason, and the flipside for say entities with something like a time based id, it will just keep piling on like a tiny memory leak...
Thoughts?

duckyuck commented on src/datoms_differ/datom.cljc:
```diff
@@ -116,39 +115,34 @@
+  (if (extends? DatomValueComparator (type v1))
+    (customCompareTo v1 v2)
+    (cmp v1 v2)))
+
```
How about dropping this function altogether, and rather supply a default implementation (i.e., using cmp) of customCompareTo for java.lang.Obect / js/object?

rundis replied:
Not sure I understand. `cmp` calls clojure `compare` which requires that the object implements `java.util.Comparable` no? 

Ahh... so that would result in an end result as if you haven't done anything, but you avoid the `extends?` check all together? Am I onto something ? :)

duckyuck replied:
Yes. Let Clojure do the polymorphic dispatching instead of you doing so via extend?.

duckyuck commented on src/datoms_differ/datom.cljc:
```diff
@@ -116,39 +115,34 @@
 
+(defprotocol DatomValueComparator
+  "A protocol that allows you to provide custom compare for types of your choosing for the value attribute of a datom"
+  (customCompareTo [this other]))
```
What's up with camelCasing, bro? :)

I'd also rename this function to "compare-value". The "custom" part is implicit in this being an protocol and thus extensible.

rundis replied:
👍

magnars commented on src/datoms_differ/datom.cljc:
```diff
@@ -0,0 +1,177 @@
+  clojure.lang.IPersistentCollection
+  (equiv [d o] (and (instance? Datom o) (equiv-datom d o)))
+  (empty [d] (throw (UnsupportedOperationException. "empty is not supported on Datom")))
+  (count [d] 5)
```
This returns `5`, but `(seq d)` returns `(list (.-e d) (.-a d) (.-v d) (.-s d))`, which is 4 elements. Am I missing something?

rundis replied:
No probably not... just a copy/paste without reflecting from https://github.com/tonsky/datascript/blob/master/src/datascript/db.cljc#L191

magnars commented on src/datoms_differ/datom.cljc:
```diff
@@ -0,0 +1,177 @@
+       (not= (.-v a) (.-v b))))
+
+(defn source-equals? [source ^Datom d]
+  (identical? source (.s d)))
```
I don't think we have a strong opinion on what constitutes a source, but this sure narrows it down a lot. Maybe we should just be up front about sources being keywords? Or do we allow other values here? In that case, `identical?` should probably be replaced by an `=`.

rundis replied:
Good catch!

`identical?` is really fast! It becomes most visible fro small changes to a large db. So for an existing db with 11000 datoms, and update resulting in maybe 100 additions went from 0.5ms to 1.1 ms (criterium quick-bench). 

I'd also add  that allowing whatever as source is a potential footgun when it comes to performance

Maybe people can live with the restriction of using keywords? Probably worth both documenting and adding a runtime check for that in `transact!` and `transact-sources!`.

magnars replied:
Restricting sources to be keywords sounds like the way to go. :)

magnars replied:
TIL strings are not interned, `(not (identical? "foobar" (str "foo" "bar")))`

magnars commented on test/datoms_differ/api_test.cljc:
```diff
@@ -0,0 +1,492 @@
+          (sut/with @db :prep [{:vessel/imo "123"}
+                               {:vessel/imo "234"}
+                               {:vessel/imo "345"}])
+          (is (= :should-throw :didnt))
```
Check out `(is (thrown-with-msg? ...))` here :) https://clojure.github.io/clojure/clojure.test-api.html

rundis replied:
👌

magnars commented on src/datoms_differ/api.cljc:
```diff
@@ -0,0 +1,281 @@
+                              (group-by :s)
+                              (map-vals #(->> % (map v-fn) set)))}))))
+
+(defn- calc-source-report [eavs source datoms]
```
Would you mind adding a docstring to this one to help with my understanding? :-)

magnars replied:
I figured it out by looking at the return value. I think this code would be clearer if it is split up into two parts: finding `old` and `new` (maybe `removed` and `added` would be clearer?) - and a separate function to update the `eavs` index with these changes.

rundis replied:
Yeah it's a mouthful, I'll split it. How about `find-source-diffs` and `update-eavs-for-source` or something like that?

magnars replied:
Excellent 👌

magnars commented:
I don't think ref-reuse is useful. Avoiding memory leaks sounds good to me 👍

magnars commented on src/datoms_differ/api.cljc:
```diff
@@ -0,0 +1,281 @@
+                                 [old new eavs] (calc-source-report (:eavs db) source datoms)]
+                             (-> db
+                                 (update :to-remove into old)
+                                 (update :to-add into new)
```
If I understand correctly, these would be better named `:removed` and `:added`? They are already removed and added in `eavs`, so these are only used to create the tx-report?

magnars replied:
Maybe even `:retracted` :)

rundis replied:
👍

magnars commented on test/datoms_differ/api_test.cljc:
```diff
@@ -0,0 +1,492 @@
+                 (is (= {:attr :vessel/name
+                         :entity-ref [:vessel/imo "123"]
+                         :conflict {:source #{"Fyken" "Syken"}}}
+                        (ex-data e)))))))))
```
Am I misunderstanding this test? A single source is allowed to change a value for an [e a]. Conflict occurs when two different sources assert different values for the same [e a].

magnars replied:
Hmm, I now see the `concat` on `entities`. But why is it that `disallow-conflicting-sources` triggers for same source?

rundis replied:
Well the name is misleading. It's a merge of what was previously `disallow-conflicting-values` and `disallow-conflicting-sources`. Maybe `disallow-conflicts` or `disallow-conflicting-values` (regardless of source implied).

Maybe this clarifies my (potentially faulty) rationale why the one in api  triggers for the same source:
1. In the previous implementation it was doing `disallow-conflicting-values` for a given set of entities for a given source. 
2. Then later on `disallow-conflicting-sources` grouped all datoms accross sources by [e a] and checked for conflicting values (ie regardless of source). That includes the  the source and it's datoms previously checked in 1 (assoced in `update-db-with-source-entity-maps`). 

So unless I'm missing something isn't 1. just forcing a potential throw error earlier (with a different  ex-info) as if you've just left it out?

Given that my assumption/reasoning holds, I made the ex-info bit generic, hopefully capturing both conflicts accross sources, or conflicts for a given source just being updated(essentially replaced in eavs that is).

magnars replied:
Renaming to `disallow-conflicting-values` seems good to me. What then about the check in `explode`? Could that be delegated to this also?

rundis replied:
explode doesn't need a source, so at the price of some duplication you get a better more targeted exception with having it just reusing the `find-conflicting-value` function was my thinking behind not reusing `disallow...`.

magnars replied:
Good call. :)

magnars reviewed (commented):
Well done on this. :-)

magnars commented on test/datoms_differ/api_test.cljc:
```diff
@@ -0,0 +1,492 @@
+               [:vessel/imo "123"] 1024}}))
+
+    (testing "empty db - nil val for attr throws"
+      (is (thrown? Exception
```
Maybe use `thrown-with-msg?` here as well? I love a good error message. :)

magnars commented on test/datoms_differ/api_test.cljc:
```diff
@@ -0,0 +1,492 @@
+          (is (= {:entity-ref [:route/number "100"]
+                  :attr :route/name
+                  :conflict {:prepare-routes #{"Stavanger-Taua" "Stavanger-Tau"}}}
+                 (ex-data e)))))))
```
Checking the ex-data is great. I don't think there's a corresponding `thrown-with-ex-data?`, so we should probably keep this construct here. :)

magnars commented on test/datoms_differ/api_test.cljc:
```diff
@@ -0,0 +1,492 @@
+          :eavs-added #{(d/datom 1025 :route/number "700" :prepare-routes)
+                        (d/datom 1025 :route/name "Døvær" :prepare-routes)}
+          :refs-added {[:route/number "700"] 1025
+                       [:route/number "800"] 1024}})))
```
If I understand the test correctly, the ref for route "800" should not be added at this point.

rundis committed:
Fixes after pr-review

- Prune diffs no longer in eavs after an update
- Refactor DatomValueComparator to provide default impl for Object and nil
- Various fixes/refactorings based on review comments

rundis commented:
Ok, hopefully addressed most of the comments. Now also cleaning up lookup-refs no longer present in the eavs index.

rundis commented:
One thing I forgot to mention: The tests that try to cover "optimized" paths in the implementation (based on ratio's it does this or that), these are kind of brittle and very implicit. At some point one might change these ratios and voila the tests no longer cover that/those paths. Might be worth investing some effort into improving at some point. Or maybe the whole select different code paths for performance thing is a dubious idea (maybe on a cloud server the performance characteristics are so different to my machine (tm) that it has a negative impact for the common scenarios for a particular users usage scenarios.
